### PR TITLE
Fix ICE with struct ctors and const generics.

### LINKED ISSUE #60818: ICE when using struct constructor with explicit const generic arguments
This very simple example ([Playground link](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=516357843078a0b1bdaef643bc20c8a5)):

```rust
#![feature(const_generics)]

struct Generic<const V: usize>;

fn main() {
    let _ = Generic::<0>;
}
```

results in: `error: internal compiler error: src/librustc_typeck/collect.rs:1407: unexpected const parent path def Def(Ctor(Struct, Const), DefId(0:13 ~ playground[d52b]::Generic[0]::{{constructor}}[0]))`. Changing it to use curly braces around `0` does not help.

### PR DESCRIPTION
**Pull request by davidtwco**
Fixes #60818.

r? @varkor

### CODE REVIEW

File: src/librustc_typeck/collect.rs [hunk 1]
```diff
@@ -1377,42 +1377,46 @@ pub fn checked_type_of<'a, 'tcx>(
...
-                            match path.res {
-                                // We've encountered an `AnonConst` in some path, so we need to
-                                // figure out which generic parameter it corresponds to and return
-                                // the relevant type.
+
+                            // We've encountered an `AnonConst` in some path, so we need to
+                            // figure out which generic parameter it corresponds to and return
+                            // the relevant type.
+                            let generics = match path.res {
+                                Res::Def(DefKind::Ctor(CtorOf::Struct, _), def_id) =>
```
Comment (varkor) [inline src/librustc_typeck/collect.rs]: This should be applicable to `CtorOf::Enum` too.

File: src/librustc_typeck/collect.rs [hunk 2]
```diff
@@ -1377,42 +1377,46 @@ pub fn checked_type_of<'a, 'tcx>(
...
-                                            }
-                                            param_index += 1;
-                                        }
-                                    }
-                                    // This is no generic parameter associated with the arg. This is
-                                    // probably from an extra arg where one is not needed.
-                                    return Some(tcx.types.err);
-                                }
-                                Res::Err => tcx.types.err,
+                                | Res::Def(DefKind::Fn, def_id) =>
```
Comment (varkor) [inline src/librustc_typeck/collect.rs]: I think we should just collapse all the above arms into `Res::Def(_, def_id) => tcx.generics_of(def_id)`: it means we won't accidentally miss any more cases in the future.

File: src/librustc_typeck/collect.rs [hunk 3]
```diff
@@ -1377,42 +1377,46 @@ pub fn checked_type_of<'a, 'tcx>(
...
                                             "unexpected const parent path def {:?}", x
                                         ),
                                     );
-                                    tcx.types.err
+                                    return Some(tcx.types.err);
                                 }
-                            }
+                            };
+
+                            generics.params.iter()
```
Comment (varkor) [inline src/librustc_typeck/collect.rs]: You can probably similarly clean up the `arg_index` calculation above if you fancy it (but no obligation to do so).

**Review by varkor — APPROVED** [review]

**Comment by davidtwco** [comment]
I've force-pushed some changes, resolving the review feedback. This will still fail CI on the expected test.

**Comment by yodaldevoid** [comment]
With #60742 merged, this should no longer be blocked.

**Comment by davidtwco** [comment]
I've pushed a version rebased atop #60742 which was merged, it should now pass CI.

**Comment by varkor** [comment]
r=me once Travis passes.

**Comment by yodaldevoid** [comment]
Travis passed, this should be ready to roll.

**Pull request merged**